### ISSUE TITLE
"object" has no attribute "__iter__" false-positive

### ISSUE THREAD

**Issue by tucked** [python/mypy#8706]
``` python
# sscce.py


def no_repro(bar: str = '4'):
    key_values = {
        'key': [1, 2, 3],
        'key2': [1, 2],
    }
    return all(
        value
        for key, values in key_values.items()
        for value in values
    )


def repro(bar: str = '4'):
    key_values = {
        'key': [1, 2, 3, bar],  # <--- bar is included
        'key2': [1, 2],
    }
    return all(
        value
        for key, values in key_values.items()
        for value in values
    )
```
- Note: no repro without `'key2': [1, 2],`
``` sh
mypy sscce.py
```
```
sscce.py:21: error: "object" has no attribute "__iter__"; maybe "__str__" or "__dir__"? (not iterable)
Found 1 error in 1 file (checked 1 source file)
```
- Python 3.6.8
- mypy 0.770 (`mypy-0.770+dev.833fc2caec6773683d691a72058293ae87139210` too)

Potentially related: #4975

**Comment by msullivan** [comment]
This is frustrating but I don't think it's a bug.

We infer the type `List[object]` for the first list and `List[int]` for the second list. `List[int]` is not a subtype of `List[object]`, so the best common subtype for them is unfortunately `object`.

**Comment by tucked** [comment]
Is the correct fix to use `tuple`s instead?

**Comment by JukkaL** [comment]
The recommended fix is to add an annotation for `key_values`, but using tuples may also work.
